bug: Graphql Assertion throws exception when validating when using Kotlin Coroutines

## Expected behavior
Validation passes and app starts as usual

## Actual behavior

Looks to have failed because of the dollar sign when the method has the suspend keyword.
Exception is thrown with this stacktrace
```
Caused by: graphql.AssertException: Name must be non-null, non-empty and match [_A-Za-z][_0-9A-Za-z]* - was 'session$suspendImpl'
at graphql.Assert.assertValidName(Assert.java:117) ~[graphql-java-17.3.jar:na]
at graphql.schema.FieldCoordinates.assertValidNames(FieldCoordinates.java:54) ~[graphql-java-17.3.jar:na]
at graphql.schema.GraphQLCodeRegistry$Builder.dataFetcher(GraphQLCodeRegistry.java:366) ~[graphql-java-17.3.jar:na]
at graphql.schema.GraphQLCodeRegistry$Builder.dataFetcher(GraphQLCodeRegistry.java:323) ~[graphql-java-17.3.jar:na]
at com.netflix.graphql.dgs.internal.DgsSchemaProvider.registerDataFetcher(DgsSchemaProvider.kt:276) ~[graphql-dgs-4.9.16.jar:4.9.16]
at com.netflix.graphql.dgs.internal.DgsSchemaProvider.findDataFetchers$lambda-12$lambda-11$lambda-10(DgsSchemaProvider.kt:209) ~[graphql-dgs-4.9.16.jar:4.9.16]
at org.springframework.core.annotation.TypeMappedAnnotations$AggregatesSpliterator.tryAdvance(TypeMappedAnnotations.java:599) ~[spring-core-5.3.15.jar:5.3.15]
at org.springframework.core.annotation.TypeMappedAnnotations$AggregatesSpliterator.tryAdvance(TypeMappedAnnotations.java:566) ~[spring-core-5.3.15.jar:5.3.15]
at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:332) ~[na:na]
at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762) ~[na:na]
at com.netflix.graphql.dgs.internal.DgsSchemaProvider.findDataFetchers(DgsSchemaProvider.kt:208) ~[graphql-dgs-4.9.16.jar:4.9.16]
at com.netflix.graphql.dgs.internal.DgsSchemaProvider.schema(DgsSchemaProvider.kt:112) ~[graphql-dgs-4.9.16.jar:4.9.16]
at com.netflix.graphql.dgs.autoconfig.DgsAutoConfiguration.schema(DgsAutoConfiguration.kt:169) ~[graphql-dgs-spring-boot-oss-autoconfigure-4.9.16.jar:4.9.16]
at com.netflix.graphql.dgs.autoconfig.DgsAutoConfiguration$$EnhancerBySpringCGLIB$$22543706.CGLIB$schema$0(<generated>) ~[graphql-dgs-spring-boot-oss-autoconfigure-4.9.16.jar:4.9.16]
at com.netflix.graphql.dgs.autoconfig.DgsAutoConfiguration$$EnhancerBySpringCGLIB$$22543706$$FastClassBySpringCGLIB$$d7467a1a.invoke(<generated>) ~[graphql-dgs-spring-boot-oss-autoconfigure-4.9.16.jar:4.9.16]
at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244) ~[spring-core-5.3.15.jar:5.3.15]
at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331) ~[spring-context-5.3.15.jar:5.3.15]
at com.netflix.graphql.dgs.autoconfig.DgsAutoConfiguration$$EnhancerBySpringCGLIB$$22543706.schema(<generated>) ~[graphql-dgs-spring-boot-oss-autoconfigure-4.9.16.jar:4.9.16]
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154) ~[spring-beans-5.3.15.jar:5.3.15]
... 89 common frames omitted
```

## Steps to reproduce
graphql.schema
```
type Query {
    session(id: String): Session
}

type Mutation {
    setSession(id: String, name: String): Boolean
}

type Session {
    id: String
    name: String
}
```
fetcher class
```
@DgsComponent
class SessionFetcher(
    private val redisOperations: ReactiveRedisOperations<String, Session>
) {

    @DgsQuery
    suspend fun session(id: String): Session? {
        return redisOperations.opsForValue().getAndAwait(id)
    }

    @DgsMutation
    suspend fun setSession(id: String, name: String): Boolean {
        return redisOperations.opsForValue().setAndAwait(id, Session(id, name))
    }
}
```

redis configuration
```
@Configuration
class RedisConfig {
    @Bean
    @Primary
    fun connectionFactory(): ReactiveRedisConnectionFactory {
        return LettuceConnectionFactory("localhost", 6379)
    }

    @Bean
    fun redisOperations(connectionFactory: LettuceConnectionFactory, objectMapper: ObjectMapper): ReactiveRedisOperations<String, Session> {
        val valueSerializer = Jackson2JsonRedisSerializer(Session::class.java).apply {
            setObjectMapper(objectMapper)
        }
        return ReactiveRedisTemplate(
            connectionFactory,
            newSerializationContext<String, Session>(StringRedisSerializer())
                .value(valueSerializer)
                .build()
        )
    }
}
```

session class
```
class Session(
    val id: String,
    val name: String
)
```

build.gradle.kts

```
import org.jetbrains.kotlin.gradle.tasks.KotlinCompile

plugins {
id("org.springframework.boot") version "2.6.3"
id("io.spring.dependency-management") version "1.0.11.RELEASE"
kotlin("jvm") version "1.6.10"
kotlin("plugin.spring") version "1.6.10"
}

group = "com.company"
version = "0.0.1-SNAPSHOT"
java.sourceCompatibility = JavaVersion.VERSION_16

repositories {
mavenCentral()
}

dependencies {
    implementation(platform("com.netflix.graphql.dgs:graphql-dgs-platform-dependencies:latest.release"))
    implementation("com.netflix.graphql.dgs:graphql-dgs-webflux-starter")
    implementation("org.jetbrains.kotlin:kotlin-reflect")
    implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8")
    implementation("org.springframework.boot:spring-boot-starter-webflux")
    implementation("org.springframework.boot:spring-boot-starter-actuator")
    implementation("org.springframework.boot:spring-boot-starter-data-redis-reactive:2.6.3")
    testImplementation("org.springframework.boot:spring-boot-starter-test")
}

tasks.withType<KotlinCompile> {
kotlinOptions {
freeCompilerArgs = listOf("-Xjsr305=strict")
jvmTarget = "16"
}
}

tasks.withType<Test> {
useJUnitPlatform()
}

```

noticed there were resources that states coroutines were supported found here
https://github.com/hantsy/spring-graphql-sample/tree/master/dgs-kotlin-co
https://github.com/Netflix/dgs-framework/issues/413

but I could not find any leads which points to an answer. This works without coroutines if I take out the suspend key words and just use  java reactors (Mono and Flux and removing the AndAwait) but using coroutines returns this exception.

 Not sure if this is entirely a bug or if I am doing something wrong. If I did something wrong, any guidance would be much appreciated

--- a/graphql-dgs/build.gradle.kts
+++ b/graphql-dgs/build.gradle.kts
@@ -39,4 +39,5 @@ dependencies {
     testImplementation("io.projectreactor:reactor-core")
     testImplementation("io.projectreactor:reactor-test")
     testImplementation("com.graphql-java:graphql-java-extended-scalars")
+    testImplementation("org.jetbrains.kotlinx:kotlinx-coroutines-test")
 }

--- a/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DataFetcherResultProcessor.kt
+++ b/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DataFetcherResultProcessor.kt
@@ -1,0 +1,32 @@
+/*
+ * Copyright 2022 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.graphql.dgs.internal
+
+import com.netflix.graphql.dgs.DgsDataFetchingEnvironment
+
+interface DataFetcherResultProcessor {
+
+    fun supportsType(originalResult: Any): Boolean
+
+    fun process(originalResult: Any, dfe: DgsDataFetchingEnvironment): Any = process(originalResult)
+
+    @Deprecated(
+        "Replaced with process(originalResult, dfe)",
+        replaceWith = ReplaceWith("process(originalResult: Any, dfe: DgsDataFetchingEnvironment)")
+    )
+    fun process(originalResult: Any): Any = originalResult
+}

--- a/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DatafetcherReference.kt
+++ b/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DatafetcherReference.kt
@@ -1,0 +1,28 @@
+/*
+ * Copyright 2022 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.graphql.dgs.internal
+
+import org.springframework.core.annotation.MergedAnnotations
+import java.lang.reflect.Method
+
+data class DatafetcherReference(
+    val instance: Any,
+    val method: Method,
+    val annotations: MergedAnnotations,
+    val parentType: String,
+    val field: String
+)

--- a/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DgsSchemaProvider.kt
+++ b/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DgsSchemaProvider.kt
@@ -79,7 +79,10 @@ class DgsSchemaProvider(
 
     private val defaultParameterNameDiscoverer = DefaultParameterNameDiscoverer()
 
-    fun schema(schema: String? = null, fieldVisibility: GraphqlFieldVisibility = DefaultGraphqlFieldVisibility.DEFAULT_FIELD_VISIBILITY): GraphQLSchema {
+    fun schema(
+        schema: String? = null,
+        fieldVisibility: GraphqlFieldVisibility = DefaultGraphqlFieldVisibility.DEFAULT_FIELD_VISIBILITY
+    ): GraphQLSchema {
         val startTime = System.currentTimeMillis()
         val dgsComponents = applicationContext.getBeansWithAnnotation(DgsComponent::class.java).values
         val hasDynamicTypeRegistry =
@@ -97,12 +100,14 @@ class DgsSchemaProvider(
             mergedRegistry = mergedRegistry.merge(existingTypeDefinitionRegistry.get())
         }
 
-        val federationResolverInstance = federationResolver.orElseGet { DefaultDgsFederationResolver(this, dataFetcherExceptionHandler) }
+        val federationResolverInstance =
+            federationResolver.orElseGet { DefaultDgsFederationResolver(this, dataFetcherExceptionHandler) }
 
         val entityFetcher = federationResolverInstance.entitiesFetcher()
         val typeResolver = federationResolverInstance.typeResolver()
         val codeRegistryBuilder = GraphQLCodeRegistry.newCodeRegistry().fieldVisibility(fieldVisibility)
-        val runtimeWiringBuilder = RuntimeWiring.newRuntimeWiring().codeRegistry(codeRegistryBuilder).fieldVisibility(fieldVisibility)
+        val runtimeWiringBuilder =
+            RuntimeWiring.newRuntimeWiring().codeRegistry(codeRegistryBuilder).fieldVisibility(fieldVisibility)
 
         dgsComponents.asSequence()
             .mapNotNull { dgsComponent -> invokeDgsTypeDefinitionRegistry(dgsComponent, mergedRegistry) }
@@ -139,7 +144,10 @@ class DgsSchemaProvider(
         }
     }
 
-    private fun invokeDgsTypeDefinitionRegistry(dgsComponent: Any, registry: TypeDefinitionRegistry): TypeDefinitionRegistry? {
+    private fun invokeDgsTypeDefinitionRegistry(
+        dgsComponent: Any,
+        registry: TypeDefinitionRegistry
+    ): TypeDefinitionRegistry? {
         return dgsComponent.javaClass.methods.asSequence()
             .filter { it.isAnnotationPresent(DgsTypeDefinitionRegistry::class.java) }
             .map { method ->
@@ -197,9 +205,10 @@ class DgsSchemaProvider(
     ) {
         dgsComponents.forEach { dgsComponent ->
             val javaClass = AopUtils.getTargetClass(dgsComponent)
-            ReflectionUtils.getUniqueDeclaredMethods(javaClass).asSequence()
+            ReflectionUtils.getUniqueDeclaredMethods(javaClass, ReflectionUtils.USER_DECLARED_METHODS).asSequence()
                 .filter { method ->
-                    MergedAnnotations.from(method, MergedAnnotations.SearchStrategy.TYPE_HIERARCHY)
+                    MergedAnnotations
+                        .from(method, MergedAnnotations.SearchStrategy.TYPE_HIERARCHY)
                         .isPresent(DgsData::class.java)
                 }
                 .forEach { method ->
@@ -308,7 +317,13 @@ class DgsSchemaProvider(
     private fun createBasicDataFetcher(method: Method, dgsComponent: Any, isSubscription: Boolean): DataFetcher<Any?> {
         return DataFetcher<Any?> { environment ->
             val dfe = DgsDataFetchingEnvironment(environment)
-            val result = DataFetcherInvoker(cookieValueResolver, defaultParameterNameDiscoverer, dfe, dgsComponent, method).invokeDataFetcher()
+            val result = DataFetcherInvoker(
+                cookieValueResolver,
+                defaultParameterNameDiscoverer,
+                dfe,
+                dgsComponent,
+                method
+            ).invokeDataFetcher()
             when {
                 isSubscription -> {
                     result
@@ -461,15 +476,3 @@ class DgsSchemaProvider(
         private val logger: Logger = LoggerFactory.getLogger(DgsSchemaProvider::class.java)
     }
 }
-
-interface DataFetcherResultProcessor {
-    fun supportsType(originalResult: Any): Boolean
-    fun process(originalResult: Any, dfe: DgsDataFetchingEnvironment): Any = process(originalResult)
-    @Deprecated(
-        "Replaced with process(originalResult, dfe)",
-        replaceWith = ReplaceWith("process(originalResult: Any, dfe: DgsDataFetchingEnvironment)")
-    )
-    fun process(originalResult: Any): Any = originalResult
-}
-
-data class DatafetcherReference(val instance: Any, val method: Method, val annotations: MergedAnnotations, val parentType: String, val field: String)

--- a/graphql-dgs/src/test/kotlin/com/netflix/graphql/dgs/CoroutineDataFetcherTest.kt
+++ b/graphql-dgs/src/test/kotlin/com/netflix/graphql/dgs/CoroutineDataFetcherTest.kt
@@ -49,25 +49,18 @@ class CoroutineDataFetcherTest {
             suspend fun concurrent(@InputArgument from: Int, to: Int): Int = coroutineScope {
                 var sum = 0
                 withContext(executor.asCoroutineDispatcher()) {
-                    println("before $from")
                     repeat(from.rangeTo(to).count()) {
                         sum++
-
                         // Forcing a blocking call to demonstrate running with a thread pool
                         Thread.sleep(50)
                     }
-                    println("after $from")
-
                     sum
                 }
             }
         }
 
         every { applicationContextMock.getBeansWithAnnotation(DgsComponent::class.java) } returns mapOf(
-            Pair(
-                "concurrentFetcher",
-                fetcher
-            )
+            "concurrentFetcher" to fetcher
         )
         every { applicationContextMock.getBeansWithAnnotation(DgsScalar::class.java) } returns emptyMap()
         every { applicationContextMock.getBeansWithAnnotation(DgsDirective::class.java) } returns emptyMap()
@@ -138,10 +131,7 @@ class CoroutineDataFetcherTest {
         }
 
         every { applicationContextMock.getBeansWithAnnotation(DgsComponent::class.java) } returns mapOf(
-            Pair(
-                "exceptionWithMessageFetcher",
-                fetcher
-            )
+            "exceptionWithMessageFetcher" to fetcher
         )
         every { applicationContextMock.getBeansWithAnnotation(DgsScalar::class.java) } returns emptyMap()
         every { applicationContextMock.getBeansWithAnnotation(DgsDirective::class.java) } returns emptyMap()

